app_history.d.ts

```diff
@@ -0,0 +1,99 @@
+interface Window {
+  readonly appHistory: AppHistory;
+}
+
+interface AppHistory extends EventTarget {
+  readonly current: AppHistoryEntry|null;
+  readonly transition: AppHistoryTransition|null;
+
+  entries(): AppHistoryEntry[];
+
+  readonly canGoBack: boolean;
+  readonly canGoForward: boolean;
+
+  navigate(url: string, options?: AppHistoryNavigateOptions): Promise<void>;
+  reload(options?: AppHistoryReloadOptions): Promise<void>;
+
+  goTo(key: string, options?: AppHistoryNavigationOptions): Promise<void>;
+  back(options?: AppHistoryNavigationOptions): Promise<void>;
+  forward(options?: AppHistoryNavigationOptions): Promise<void>;
+
+  onnavigate: ((ev: Event) => any)|null;
+  onnavigatesuccess: ((ev: Event) => any)|null;
+  onnavigateerror: ((ev: Event) => any)|null;
+}
+
+interface AppHistoryTransition {
+  readonly navigationType: AppHistoryNavigationType;
+  readonly from: AppHistoryEntry;
+  readonly finished: Promise<void>;
+
+  rollback(options?: AppHistoryNavigationOptions): Promise<void>;
+}
+
+interface AppHistoryEntry extends EventTarget {
+  readonly key: string;
+  readonly id: string;
+  readonly url: string;
+  readonly index: number;
+  readonly sameDocument: boolean;
+
+  getState(): unknown;
+
+  onnavigateto: ((ev: Event) => any)|null;
+  onnavigatefrom: ((ev: Event) => any)|null;
+  onfinish: ((ev: Event) => any)|null;
+  ondispose: ((ev: Event) => any)|null;
+}
+
+type AppHistoryNavigationType = 'reload'|'push'|'replace'|'traverse';
+
+interface AppHistoryNavigationOptions {
+  navigateInfo?: unknown;
+}
+
+interface AppHistoryNavigateOptions extends AppHistoryNavigationOptions {
+  state?: unknown;
+  replace?: boolean;
+}
+
+interface AppHistoryReloadOptions extends AppHistoryNavigationOptions {
+  state?: unknown;
+}
+
+declare class AppHistoryNavigateEvent extends Event {
```

Comment:
When do you choose `declare class` vs. `interface`? On the spec side everything is a class, but most of them have constructors that throw if you try to use them.

dom.d.ts seems to do something strange for most classes where they do e.g. `interface HTMLBaseElement { ... }` and then

```js
declare var HTMLBaseElement: {
    prototype: HTMLBaseElement;
    new(): HTMLBaseElement;
};
```

which doesn't really make sense to me...

/cc @saschanaz who I believe helps maintain dom.d.ts.

Reply:
That's [a decision from TypeScript team](https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/858#issuecomment-644294982) that [I don't exactly follow](https://github.com/microsoft/TypeScript/issues/39054#issuecomment-644299696), but one of the reasons was that the `interface` pattern is easier to extend than a class.

/cc @sandersn

Reply:
Wow, yeah, that's bizarre that they think the DOM classes are not classes. I guess that makes me lean toward using `declare class` for all of them?

What's the difference between just `interface` and the `interface` + `declare var` pattern?

Reply:
Just `interface` does not allow reference to HTMLBaseElement e.g. `instanceof HTMLBaseElement` since it's not a value but a type. (See [demo](https://www.typescriptlang.org/play?ssl=11&ssc=28&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgGIHt0FEA2EC2E4yA3gL4CwAUNQBTnKgDOYcIS6MamuBRYASgDc1aqEixEKAEJwovQsXLUAJhAQ45KAG5zks+XkVgAXKWrJLyAA5R0YewE9rEMwYX8RVK8hAQA7rQCbnIe4F6UNFT0ZIwgLGwcXO5G-MJAA))

Reply:
>I guess that makes me lean toward using declare class for all of them?

Yeah, it shouldn't harm here.

Reply:
I'll go with declare class

Reply:
Briefly, if a class needs to be extensible via merging (declaring twice), you can't use `declare class` because it's block-scoped and doesn't allow multiple declarations. You have to use the weird structure like for HTMLBaseElement.

In other words, this doesn't work:

```ts
// @filename: app_history.d.ts
declare class AppHistoryNavigateEvent extends Event {
  // properties from this PR
}

// @filename: somebody_else_extends_app_history.d.ts
declare class AppHistoryNavigateEvent extends Event {
  // properties that somebody else has added later
}
```

Reply:
If users are supposed to be able to tack properties onto instances of AppHistoryNavigateEvent, these should be written using the decomposed class pattern so that other d.ts files can declare those properties.

Reply:
I don't think this expects any extension. And classes are still extendable by interfaces, no?

Reply:
> And classes are still extendable by interfaces, no?

Yes. The example code I gave is trying to provide typing for people adding expando properties to `AppHistoryNavigateEvent` instances, not people using `X extends AppHistoryNavigateEvent`.

Reply:
Not sure I follow, I mean [declaration merging](https://www.typescriptlang.org/play?#code/MYGwhgzhAECCAO8ASBLCAXA9gJwJ4DkwA3FAczHQFMBRIygO3WgG8BfAbgFgAoHlRytgBmYYJTiJUGHAWJkKNOoxY9oa6JQAe8MPQAmmAArZM8QelwAuaPQCuAWwBGgrt1Y8e9SgHcJyNFh4hCTkVLQM6AAUAJQAdFo6+kYmZtgW7EA).

Reply:
Yep, that's what I'm talking about too. Declaration merging allows types to represent expando properties on values.

Reply:
Yeah and that doesn't require decomposed class pattern, unless someone wants to add static properties (which doesn't work anyway in decomposed class pattern, unless there is no named interface for the constructor part)?

Reply:
Representing expando properties does require a decomposed class pattern + declaration merging as far as I know. Is there another way?

Reply:
The following works as I said in https://github.com/WICG/app-history/pull/140#discussion_r679560568...

```ts
class AppHistoryNavigateEvent {};

interface AppHistoryNavigateEvent {
    expandoProperty: number;
}

new AppHistoryNavigateEvent().expandoProperty;
```

Reply:
Sorry, I didn't follow the link and assumed you meant merging two interfaces with the same name instead of merging a class and an interface.

That pattern is much easier to read, and the only thing it misses is the ability to polyfill by copying the entire declaration.